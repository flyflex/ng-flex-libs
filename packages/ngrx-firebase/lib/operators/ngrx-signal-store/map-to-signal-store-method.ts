import { switchMap } from 'rxjs/operators';

import { ActionOptions } from '../../models/action-options.model';
import { SignalStoreMethodsMap } from '../../models/action-types.model';
import { StoreAction } from '../../models/store-action.model';
import { SignalStoreMethod } from '../../models/signal-store-method.model';

/**
 * Map grouped firebase document actions to NGRX signal store method
 */

export const mapToSignalStoreMethod = <T>(
  methodsMap: SignalStoreMethodsMap,
  {
    parentId,
    includeParentIdInPayload,
    parentIdPayloadKey,
    includeParentIdInNoResults,
    parentIdProp,
    payloadProp,
  }: Partial<ActionOptions<T>> = {},
) =>
  switchMap((groupedActions: StoreAction<T>[]) =>
    groupedActions.map((storeAction: StoreAction<T>) => {
      if (storeAction.actionName === 'loadNoResults') {
        if (includeParentIdInNoResults && parentId) {
          return {
            methodName: methodsMap[storeAction.actionName],
            payload: {
              [parentIdProp]: parentId,
            }
          } as SignalStoreMethod<T>;
        } else {
          return {
            methodName: methodsMap[storeAction.actionName],
          } as SignalStoreMethod<T>;
        }
      }

      const payload = Array.isArray(storeAction.payload) && includeParentIdInPayload && parentIdPayloadKey
        ? storeAction.payload.map(entity => ({
          ...entity,
          [parentIdPayloadKey]: parentId,
        }))
        : storeAction.payload;

      return {
        methodName: methodsMap[storeAction.actionName],
        payload: {
          [payloadProp]: payload,
          [parentIdProp]: parentId,
        }
      } as SignalStoreMethod<T>;
    }),
  );
