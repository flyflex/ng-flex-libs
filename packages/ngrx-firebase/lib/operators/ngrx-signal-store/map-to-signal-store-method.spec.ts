import { of } from 'rxjs';

import { SignalStoreMethodsMap } from '../../models/action-types.model';
import { JestUtils } from '../../tests/jest-utils';
import { mapToSignalStoreMethod } from './map-to-signal-store-method';

const actionsMocksCreators: SignalStoreMethodsMap = {
  loadNoResults: 'loadNoResultsMock',
  addMany: 'entityAddMock',
  upsertMany: 'entityUpdateMock',
  removeMany: 'entityRemoveMock',
};

const action1 = { type: 'added', payload: { id: 'id1', data: 'value1' } };
const action2 = { type: 'added', payload: { id: 'id2', data: 'value2' } };
const action3 = { type: 'removed', payload: { id: 'id3', data: 'value3' } };
const action4 = { type: 'modified', payload: { id: 'id2', data: 'changedValue2' } };
const GROUPED_ACTIONS = [
  {
    actionName: 'addMany',
    payload: [action1.payload, action2.payload],
  },
  {
    actionName: 'removeMany',
    payload: [action3.payload],
  },
  {
    actionName: 'upsertMany',
    payload: [action4.payload],
  },
];

const PARENT_ID = 'someParentId1';

describe('mapToSignalStoreMethod', () => {
  let spy;

  beforeEach(() => {
    spy = jest.fn();
    spy.mockClear();
  });

  describe(`when using signal store method creators from NGRX`, () => {
    describe('when mapping loadNoResults action', () => {
      describe('with no parent id', () => {
        it('should return the loadNoResults without parent id', () => {
          const firebaseActions = [{ actionName: 'loadNoResults' }];

          of(firebaseActions).pipe(
            mapToSignalStoreMethod(actionsMocksCreators)
          ).subscribe(spy);

          expect(spy).toHaveBeenLastCalledWith({ methodName: actionsMocksCreators.loadNoResults });
        });
      });

      describe('when including an existing parent ID through params', () => {
        it('should return the loadNoResults with matching parent id', () => {
          const firebaseActions = [{ actionName: 'loadNoResults' }];

          of(firebaseActions).pipe(
            mapToSignalStoreMethod(
              actionsMocksCreators,
              {
                parentId: PARENT_ID,
                includeParentIdInPayload: false,
                parentIdPayloadKey: null,
                includeParentIdInNoResults: true,
                parentIdProp: 'parentId',
                payloadProp: 'payloadProp',
              }
            ),
          ).subscribe(spy);

          expect(spy).toHaveBeenLastCalledWith({ methodName: actionsMocksCreators.loadNoResults, payload: { parentId: PARENT_ID } });
        });
      });

      describe('when including an existing parent ID through params', () => {
        it('should return the loadNoResults with matching parent id key', () => {
          const firebaseActions = [{ actionName: 'loadNoResults' }];

          of(firebaseActions).pipe(
            mapToSignalStoreMethod(
              actionsMocksCreators,
              {
                parentId: PARENT_ID,
                includeParentIdInPayload: false,
                parentIdPayloadKey: null,
                includeParentIdInNoResults: true,
                parentIdProp: 'parentId2',
                payloadProp: 'payloadProp'
              }
            ),
          ).subscribe(spy);

          expect(spy).toHaveBeenCalledWith({ methodName: actionsMocksCreators.loadNoResults, payload: { parentId2: PARENT_ID } });
        });
      });

      describe('when there is no parent id', () => {
        it('should return the loadNoResults without any parent id', () => {
          const firebaseActions = [{ actionName: 'loadNoResults' }];

          of(firebaseActions).pipe(
            mapToSignalStoreMethod(
              actionsMocksCreators,
              {
                parentId: null,
                includeParentIdInPayload: false,
                parentIdPayloadKey: null,
                includeParentIdInNoResults: true,
                parentIdProp: 'parentId',
                payloadProp: 'payloadProp',
              }
            ),
          ).subscribe(spy);

          expect(spy).toHaveBeenLastCalledWith({ methodName: actionsMocksCreators.loadNoResults });
        });
      });

      describe('when the option to include parent ID is set to false', () => {
        it('should return the loadNoResults with forced parent id', () => {
          const firebaseActions = [{ actionName: 'loadNoResults' }];

          of(firebaseActions).pipe(
            mapToSignalStoreMethod(
              actionsMocksCreators,
              {
                parentId: PARENT_ID,
                includeParentIdInPayload: false,
                parentIdPayloadKey: null,
                includeParentIdInNoResults: false,
                parentIdProp: 'parentIdKey',
                payloadProp: 'payloadProp',
              }
            ),
          ).subscribe(spy);

          expect(spy).toHaveBeenLastCalledWith({ methodName: actionsMocksCreators.loadNoResults });
        });
      });
    });

    describe('when mapping standard actions', () => {
      describe('by default', () => {
        describe('with parent Id', () => {
          it('should return methods with proper payload and parent id', () => {
            of(GROUPED_ACTIONS).pipe(
              mapToSignalStoreMethod(
                actionsMocksCreators,
                {
                  parentId: PARENT_ID,
                  parentIdProp: 'parentIdKey',
                  payloadProp: 'payloadProp',
                }
              ),
            ).subscribe(spy);

            expect(spy.mock.calls).toMatchSnapshot();
          });
        });

        describe('with no parent id', () => {
          it('should return methods with proper payload and empty parent id', () => {
            of(GROUPED_ACTIONS).pipe(
              mapToSignalStoreMethod(
                actionsMocksCreators,
                {
                  parentIdProp: 'parentIdKey',
                  payloadProp: 'payloadProp',
                }
              ),
            ).subscribe(spy);

            expect(spy.mock.calls).toMatchSnapshot();
          });
        });
      });

      describe('with option to include parent id in payload', () => {
        describe('when a parent id payload key is set in option', () => {
          it('should return methods with payload and parent id as given key value', () => {
            const parentKey = 'parentKey';

            of(GROUPED_ACTIONS).pipe(
              mapToSignalStoreMethod(
                actionsMocksCreators,
                {
                  parentId: PARENT_ID,
                  includeParentIdInPayload: true,
                  parentIdPayloadKey: parentKey,
                  parentIdProp: 'parentIdKey',
                  payloadProp: 'payloadProp',
                }
              ),
            ).subscribe(spy);

            expect(spy.mock.calls).toMatchSnapshot();
          });
        });

        describe('when no parent id payload key is not given in options', () => {
          it('should return methods with payload but no parent id', () => {
            of(GROUPED_ACTIONS).pipe(
              mapToSignalStoreMethod(
                actionsMocksCreators,
                {
                  parentId: PARENT_ID,
                  includeParentIdInPayload: true,
                  parentIdProp: 'parentIdKey',
                  payloadProp: 'payloadProp',
                }
              ),
            ).subscribe(spy);

            expect(spy.mock.calls).toMatchSnapshot();
          });
        });
      });
    });
  });
});
