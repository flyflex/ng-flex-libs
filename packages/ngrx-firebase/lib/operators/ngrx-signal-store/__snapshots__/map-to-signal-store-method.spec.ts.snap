// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapToSignalStoreMethod when using signal store method creators from NGRX when mapping standard actions by default with no parent id should return methods with proper payload and empty parent id 1`] = `
Array [
  Array [
    Object {
      "methodName": "entityAddMock",
      "payload": Object {
        "parentIdKey": undefined,
        "payloadProp": Array [
          Object {
            "data": "value1",
            "id": "id1",
          },
          Object {
            "data": "value2",
            "id": "id2",
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "methodName": "entityRemoveMock",
      "payload": Object {
        "parentIdKey": undefined,
        "payloadProp": Array [
          Object {
            "data": "value3",
            "id": "id3",
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "methodName": "entityUpdateMock",
      "payload": Object {
        "parentIdKey": undefined,
        "payloadProp": Array [
          Object {
            "data": "changedValue2",
            "id": "id2",
          },
        ],
      },
    },
  ],
]
`;

exports[`mapToSignalStoreMethod when using signal store method creators from NGRX when mapping standard actions by default with parent Id should return methods with proper payload and parent id 1`] = `
Array [
  Array [
    Object {
      "methodName": "entityAddMock",
      "payload": Object {
        "parentIdKey": "someParentId1",
        "payloadProp": Array [
          Object {
            "data": "value1",
            "id": "id1",
          },
          Object {
            "data": "value2",
            "id": "id2",
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "methodName": "entityRemoveMock",
      "payload": Object {
        "parentIdKey": "someParentId1",
        "payloadProp": Array [
          Object {
            "data": "value3",
            "id": "id3",
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "methodName": "entityUpdateMock",
      "payload": Object {
        "parentIdKey": "someParentId1",
        "payloadProp": Array [
          Object {
            "data": "changedValue2",
            "id": "id2",
          },
        ],
      },
    },
  ],
]
`;

exports[`mapToSignalStoreMethod when using signal store method creators from NGRX when mapping standard actions with option to include parent id in payload when a parent id payload key is set in option should return methods with payload and parent id as given key value 1`] = `
Array [
  Array [
    Object {
      "methodName": "entityAddMock",
      "payload": Object {
        "parentIdKey": "someParentId1",
        "payloadProp": Array [
          Object {
            "data": "value1",
            "id": "id1",
            "parentKey": "someParentId1",
          },
          Object {
            "data": "value2",
            "id": "id2",
            "parentKey": "someParentId1",
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "methodName": "entityRemoveMock",
      "payload": Object {
        "parentIdKey": "someParentId1",
        "payloadProp": Array [
          Object {
            "data": "value3",
            "id": "id3",
            "parentKey": "someParentId1",
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "methodName": "entityUpdateMock",
      "payload": Object {
        "parentIdKey": "someParentId1",
        "payloadProp": Array [
          Object {
            "data": "changedValue2",
            "id": "id2",
            "parentKey": "someParentId1",
          },
        ],
      },
    },
  ],
]
`;

exports[`mapToSignalStoreMethod when using signal store method creators from NGRX when mapping standard actions with option to include parent id in payload when no parent id payload key is not given in options should return methods with payload but no parent id 1`] = `
Array [
  Array [
    Object {
      "methodName": "entityAddMock",
      "payload": Object {
        "parentIdKey": "someParentId1",
        "payloadProp": Array [
          Object {
            "data": "value1",
            "id": "id1",
          },
          Object {
            "data": "value2",
            "id": "id2",
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "methodName": "entityRemoveMock",
      "payload": Object {
        "parentIdKey": "someParentId1",
        "payloadProp": Array [
          Object {
            "data": "value3",
            "id": "id3",
          },
        ],
      },
    },
  ],
  Array [
    Object {
      "methodName": "entityUpdateMock",
      "payload": Object {
        "parentIdKey": "someParentId1",
        "payloadProp": Array [
          Object {
            "data": "changedValue2",
            "id": "id2",
          },
        ],
      },
    },
  ],
]
`;
